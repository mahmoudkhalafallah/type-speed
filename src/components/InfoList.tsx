import React, { useEffect, useState, useCallback } from 'react'
import styled from 'styled-components'
import { HISTORY_URL } from '../constants'

const StyledInfoList = styled.ul`
align-self: flex-end;
text-align: end;
margin: 25px 0;
font-weight: bold;
list-style: none;
`
const StyledInfoItem = styled.li`
margin-bottom: 5px;
`

interface Props {
    endGame: React.Dispatch<React.SetStateAction<boolean>>;
    gameOver: boolean;
    wpm: number;
    cp: number;
    historyId: string
}

const InfoList: React.FC<Props> = ({ endGame, gameOver, wpm, cp, historyId }) => {
    const [time, setTime] = useState({ minutes: 0, seconds: 5 })

    const stopGame = useCallback((timerHandle: NodeJS.Timeout) => {
        clearInterval(timerHandle)
        endGame(true)
    }, [endGame])


    useEffect(() => {
        const timerHandle = setInterval(() => {
            if (time.seconds > 0) {
                setTime(t => ({ minutes: t.minutes, seconds: t.seconds - 1 }))
            } else if (time.seconds === 0) {
                if (time.minutes === 0) {
                    stopGame(timerHandle)
                    return
                }
                setTime(t => ({ minutes: t.minutes - 1, seconds: 59 }))
            }
        }, 1000)

        if (gameOver) {
            stopGame(timerHandle)
            return
        }

        return () => {
            clearInterval(timerHandle)
        }
    }, [gameOver, stopGame, time.minutes, time.seconds])

    return <StyledInfoList>
        <StyledInfoItem>Time: {time.minutes}:{time.seconds > 9 ? time.seconds : `0${time.seconds}`}</StyledInfoItem>
        <StyledInfoItem>
            <span>{wpm}</span> WPM
		</StyledInfoItem>
        {gameOver && <StyledInfoItem>
            Completion Percentage: <span>{cp}</span>%
		</StyledInfoItem>}
        {(historyId.length > 0) && <StyledInfoItem>Find your record <a href={`${HISTORY_URL}/${historyId}`} target='_blank' rel="noopener noreferrer" >here</a></StyledInfoItem>}
    </StyledInfoList>
}

export default InfoList
