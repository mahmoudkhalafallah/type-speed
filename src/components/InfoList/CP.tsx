import React from 'react'
const CP: React.FC<{ cp: number }> = ({ cp }) =>
    <>
        Completion Percentage:
        <span>{cp}%</span>
    </>
export default CP