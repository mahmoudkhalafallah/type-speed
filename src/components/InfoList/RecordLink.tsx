import React from 'react'
import { Link } from '@reach/router'
const RecordLink: React.FC<{ historyId: string }> = ({ historyId }) =>
    <>
        Find your record <Link to={`/history/${historyId}`}>here</Link>
    </>
export default RecordLink