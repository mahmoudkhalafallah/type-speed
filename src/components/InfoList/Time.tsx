import React from 'react'
const Time: React.FC<{ minutes: number, seconds: number }> = ({ minutes, seconds }) =>
    <>
        Time:
        <span>{minutes}:{seconds > 9 ? seconds : `0${seconds}`}</span>
    </>
export default Time