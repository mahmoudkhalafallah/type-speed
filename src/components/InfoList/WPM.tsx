import React from 'react'
import styled from 'styled-components'

const WpmIndicator = styled.span`
    color: ${({ wpm }: { wpm: number }) => {
        if (wpm > 35) {
            return 'green'
        } else if (wpm <= 35 && wpm > 23) {
            return '#de5615'
        } else {
            return 'red'
        }
    }};
`

const WPM: React.FC<{ wpm: number }> = ({ wpm }) =>
    <>
        Words Per Minute: <WpmIndicator wpm={wpm}>{wpm} WPM</WpmIndicator>
    </>
export default WPM