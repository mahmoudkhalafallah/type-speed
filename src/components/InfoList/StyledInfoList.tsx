import React from 'react'
import styled from 'styled-components'
import Time from './Time'
import WPM from './WPM'
import CP from './CP'
import RecordLink from './RecordLink'

const StyledList = styled.ul`
align-self: flex-end;
text-align: end;
margin: 25px 0;
list-style: none;
font-family: "capture_it";
`
const StyledItem = styled.li`
margin-bottom: 5px;
span {
    min-width: 70px;
    display: inline-block;
}
a {
    color: #3d3d3d;
}
`

interface Props {
    time: { minutes: number, seconds: number };
    gameOver: boolean;
    wpm: number;
    cp: number;
    historyId: string
}

const StyledInfoList: React.FC<Props> = ({ time, gameOver, wpm, cp, historyId }) => {
    return <StyledList>
        <StyledItem>
            <Time minutes={time.minutes} seconds={time.seconds} />
        </StyledItem>
        <StyledItem>
            <WPM wpm={wpm} />
        </StyledItem>
        {gameOver && <StyledItem>
            <CP cp={cp} />
        </StyledItem>}
        {(historyId.length > 0) && <StyledItem>
            <RecordLink historyId={historyId} />
        </StyledItem>}
    </StyledList>
}

export default StyledInfoList