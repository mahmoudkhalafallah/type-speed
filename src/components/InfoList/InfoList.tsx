import React, { useEffect, useState, useCallback } from 'react'
import StyledInfoList from './StyledInfoList'

interface Props {
    endGame: React.Dispatch<React.SetStateAction<boolean>>;
    gameOver: boolean;
}

const InfoList: React.FC<Props> = ({ endGame, gameOver }) => {
    const [time, setTime] = useState({ minutes: 3, seconds: 0 })

    const stopGame = useCallback((timerHandle: NodeJS.Timeout) => {
        clearInterval(timerHandle)
        endGame(true)
    }, [endGame])

    useEffect(() => {
        const timerHandle = setInterval(() => {
            if (time.seconds > 0) {
                setTime(t => ({ minutes: t.minutes, seconds: t.seconds - 1 }))
            } else if (time.seconds === 0) {
                if (time.minutes === 0) {
                    stopGame(timerHandle)
                    return
                }
                setTime(t => ({ minutes: t.minutes - 1, seconds: 59 }))
            }
        }, 1000)

        if (gameOver) {
            stopGame(timerHandle)
            return
        }

        return () => {
            clearInterval(timerHandle)
        }
    }, [gameOver, stopGame, time.minutes, time.seconds])

    return <StyledInfoList time={time} showCp={gameOver} />
}

export default InfoList
