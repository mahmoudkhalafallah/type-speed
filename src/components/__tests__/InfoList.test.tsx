import React from 'react'
import InfoList from '../InfoList'
import renderer from 'react-test-renderer'

const cp = 0
const wpm = 30
const historyId = ''
const gameOver = false
const endGame = () => { }

test('InfoList rendered correctly', () => {
    const component = renderer.create(
        <InfoList
            cp={cp}
            wpm={wpm}
            historyId={historyId}
            gameOver={gameOver}
            endGame={endGame}
        />,
    )
    let tree = component.toJSON()

    expect(tree).toMatchSnapshot()
})

test('InfoList rendered correctly on game over', () => {
    const component = renderer.create(
        <InfoList
            cp={20}
            wpm={wpm}
            historyId={'xyz'}
            gameOver={!gameOver}
            endGame={endGame}
        />,
    )
    let tree = component.toJSON()

    expect(tree).toMatchSnapshot()
})
