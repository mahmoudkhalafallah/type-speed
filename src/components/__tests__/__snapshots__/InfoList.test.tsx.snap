// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoList rendered correctly 1`] = `
<ul
  className="sc-bdVaJa cSScn"
>
  <li
    className="sc-bwzfXH fwsDhg"
  >
    Time:
    <span>
      3
      :
      00
    </span>
  </li>
  <li
    className="sc-bwzfXH fwsDhg"
  >
    Words Per Minute:
    <span
      className="sc-htpNat eqnzgt"
    >
      30
       WPM
    </span>
  </li>
</ul>
`;

exports[`InfoList rendered correctly on game over 1`] = `
<ul
  className="sc-bdVaJa cSScn"
>
  <li
    className="sc-bwzfXH fwsDhg"
  >
    Time:
    <span>
      3
      :
      00
    </span>
  </li>
  <li
    className="sc-bwzfXH fwsDhg"
  >
    Words Per Minute:
    <span
      className="sc-htpNat eqnzgt"
    >
      30
       WPM
    </span>
  </li>
  <li
    className="sc-bwzfXH fwsDhg"
  >
    Completion Percentage:
    <span>
      20
      %
    </span>
  </li>
  <li
    className="sc-bwzfXH fwsDhg"
  >
    Find your record 
    <a
      href="/history/xyz"
      onClick={[Function]}
    >
      here
    </a>
  </li>
</ul>
`;
